src/pal/src/map/map.cpp

```diff
@@ -2241,7 +2241,9 @@ MAPmmapAndRecord(
     PAL_ERROR palError = NO_ERROR;
     LPVOID pvBaseAddress = NULL;
 
-    pvBaseAddress = mmap(addr, len, prot, flags, fd, offset);
+    off_t adjust = offset & (VIRTUAL_PAGE_MASK);
+
+    pvBaseAddress = mmap(static_cast<char *>(addr) - adjust, len + adjust, prot, flags, fd, offset - adjust);
```

Comment:
@janvorli  Regarding suggested security hole.  The mmap of the two pages with different security permissions should not create the security hole you describe.  It will create two mappings.  Probably even to the same physical memory initially, but the "writable" one will be marked for copy on write so that it cannot modify the clone.  This is because we asked for the mapping to be PRIVATE.